back/src/Guide/Api/Types.hs

```diff
@@ -644,6 +642,42 @@ toCHeading h = CHeading
   , chSlug    = H $ headingSlug h
   }
 
+-- | Custom 'CForest' is the same as 'Forest'.
+-- | It needed to customize JSON instances for TOC to be list of object, not list of lists.
+type CForest a = [CTree a]
+
+-- | Custom 'CTree' is the same as 'Tree'.
+-- | It needed to customize JSON instances for TOC to be list of object, not list of lists.
+data CTree a = CNode
+  { cRootLabel :: a
+  , cSubForest :: CForest a
+  } deriving (Show, Generic)
+
+instance A.ToJSON (CTree CHeading) where
+  toJSON (CNode CHeading{..} cSubForest) = A.object
+    [ "content" A..= chContent
+    , "slug"    A..= chSlug
+    , "forest"  A..= cSubForest
+    ]
+
+instance A.FromJSON (CTree CHeading) where
+  parseJSON = A.withObject "CTree" $ \o -> do
+    chContent  <- o A..: "content"
+    chSlug     <- o A..: "slug"
+    cSubForest <- o A..: "forest"
+    let cRootLabel = CHeading{..}
+    pure CNode{..}
+
+instance ToSchema a => ToSchema (CTree a) where
+  declareNamedSchema = genericDeclareNamedSchema schemaOptions
```

Comment:
Pretty sure this will not give you the right Swagger model.

I would recommend changing the types to something like this:

```haskell
data CTocHeading = CTocHeading
  { cthContent     :: CMarkdown
  , cthSlug        :: Text           ? "In-page anchor for linking"
  , cthSubheadings :: [CTocHeading]
  }
```

Reply:
(And then the generic model will be the right one.)